### ISSUE TITLE
The use of target domain gt_boxes

### ISSUE THREAD

**Issue by 1061136002** [ChrisAllenMing/GPA-detection#11]
Thanks for your share!
When i study your code, I found that you used the true gt_boxes of target domain at the training stage. This is unsupervised domain adaptation,  I think you can't get the ground truth of the target domain. Is there a problem with this part？ 
Thanks for your reply!

**Comment by ChrisAllenMing** [comment]
Hi, thanks for your remind!
For evaluating the performance of the model trained with target data during development, we retain the codes of tgt_gt_boxes. Whereas, in the loss definition of the GPA model, only the supervisory signals on source domain are used, and the supervised losses for target domain are not taken, as [this line of code](https://github.com/ChrisAllenMing/GPA-detection/blob/d63d3cd8a76523615c288cceb676f75967f5914e/train_GPA.py#L491).

**Comment by 1061136002** [comment]
> Hi, thanks for your remind!
> For evaluating the performance of the model trained with target data during development, we retain the codes of tgt_gt_boxes. Whereas, in the loss definition of the GPA model, only the supervisory signals on source domain are used, and the supervised losses for target domain are not taken, as [this line of code](https://github.com/ChrisAllenMing/GPA-detection/blob/d63d3cd8a76523615c288cceb676f75967f5914e/train_GPA.py#L491).

**Comment by 1061136002** [comment]
> Hi, thanks for your remind!
> For evaluating the performance of the model trained with target data during development, we retain the codes of tgt_gt_boxes. Whereas, in the loss definition of the GPA model, only the supervisory signals on source domain are used, and the supervised losses for target domain are not taken, as [this line of code](https://github.com/ChrisAllenMing/GPA-detection/blob/d63d3cd8a76523615c288cceb676f75967f5914e/train_GPA.py#L491).

Thanks for your

**Comment by 1061136002** [comment]
> Hi, thanks for your remind!
> For evaluating the performance of the model trained with target data during development, we retain the codes of tgt_gt_boxes. Whereas, in the loss definition of the GPA model, only the supervisory signals on source domain are used, and the supervised losses for target domain are not taken, as [this line of code](https://github.com/ChrisAllenMing/GPA-detection/blob/d63d3cd8a76523615c288cceb676f75967f5914e/train_GPA.py#L491).

Thanks for your reply!
But I found that your total loss: 
loss = rpn_loss_cls.mean() + rpn_loss_box.mean() + RCNN_loss_cls.mean() + RCNN_loss_bbox.mean()  + args.da_weight * (RCNN_loss_intra + RCNN_loss_inter)  + rpn_da_weight * (RPN_loss_intra + RPN_loss_inter).
To obtain both loss_intra and loss_inter , need using variable tgt_adapt_feat， tgt_cls_prob and tgt_rois. So you have used ground truth of the target domain when computing loss.

**Comment by ChrisAllenMing** [comment]
Ok, I will check these data flows after this busy period.

**Comment by mmmll23** [comment]
Thanks for your share! I found you have used ground truth of the target domain in the training codes, such as the procedure of obtaining   variable tgt_rois_select , tgt_rois.
1.tgt_rois, tgt_rpn_loss_cls, tgt_rpn_loss_bbox, tgt_rpn_cls_prob, tgt_rois_select = self.RCNN_rpn(tgt_base_feat,
                                                                            tgt_im_info, tgt_gt_boxes, tgt_num_boxes)
2.tgt_roi_data = self.RCNN_proposal_target(tgt_rois, tgt_gt_boxes, tgt_num_boxes)
            tgt_rois, tgt_rois_label, tgt_rois_target, tgt_rois_inside_ws, tgt_rois_outside_ws = tgt_roi_data

Could you check this part of codes?  Thank you.

**Comment by neineimeow** [comment]
Hi~ I found you have used ground truth target domain boxes annotations to obtain target region proposals: 
https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/adaptive_faster_rcnn/faster_rcnn.py#L66

**Comment by drickzsy** [comment]
> > Hi, thanks for your remind!
> > For evaluating the performance of the model trained with target data during development, we retain the codes of tgt_gt_boxes. Whereas, in the loss definition of the GPA model, only the supervisory signals on source domain are used, and the supervised losses for target domain are not taken, as [this line of code](https://github.com/ChrisAllenMing/GPA-detection/blob/d63d3cd8a76523615c288cceb676f75967f5914e/train_GPA.py#L491).
> 
> Thanks for your reply!
> But I found that your total loss:
> loss = rpn_loss_cls.mean() + rpn_loss_box.mean() + RCNN_loss_cls.mean() + RCNN_loss_bbox.mean() + args.da_weight * (RCNN_loss_intra + RCNN_loss_inter) + rpn_da_weight * (RPN_loss_intra + RPN_loss_inter).
> To obtain both loss_intra and loss_inter , need using variable tgt_adapt_feat， tgt_cls_prob and tgt_rois. So you have used ground truth of the target domain when computing loss.

Hello! Sorry to bother you. Have you understand this question?

**Comment by Shuntw6096** [comment]
I found that target bounding boxes are used to calculate overlap(iou) to select fg feature and bg feature in [RPN](https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/rpn/rpn.py) ,
it is a kind of weird in unsupervised domain adaptation, or have I misunderstood something?
https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/adaptive_faster_rcnn/faster_rcnn.py#L78
https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L98
https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L112-L113
https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/proposal_target_layer_cascade.py#L50-L52

Thanks for your reply.

**Comment by ygzhu** [comment]
> I found that target bounding boxes are used to calculate overlap(iou) to select fg feature and bg feature in [RPN](https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/rpn/rpn.py) , it is a kind of weird in unsupervised domain adaptation, or have I misunderstood something?
> 
> https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/adaptive_faster_rcnn/faster_rcnn.py#L78
> 
> 
> https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L98
> 
> 
> https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L112-L113
> 
> 
> https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/proposal_target_layer_cascade.py#L50-L52
> 
> Thanks for your reply.

I agree with you. The gts of the target data are used during training process. Only the related losses of RPN and RCNN with target domain are discarded when computing the total loss. Something seems inconsistent with the unsupervised domain adaptation object detection declared in the original paper. 

Besides, if the gts of the target domain  are unavailable when training, i wonder the generation of the prorotype in target domain.

**Comment by drickzsy** [comment]
> > I found that target bounding boxes are used to calculate overlap(iou) to select fg feature and bg feature in [RPN](https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/rpn/rpn.py) , it is a kind of weird in unsupervised domain adaptation, or have I misunderstood something?
> > https://github.com/ChrisAllenMing/GPA-detection/blob/b2cff022cfecda5178e6a0efefa79adfcad6a37b/lib/model/adaptive_faster_rcnn/faster_rcnn.py#L78
> > 
> > https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L98
> > 
> > https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/anchor_target_layer.py#L112-L113
> > 
> > https://github.com/ChrisAllenMing/GPA-detection/blob/fedcb501558b6aff8eccf09d319e9712c5473bad/lib/model/rpn/proposal_target_layer_cascade.py#L50-L52
> > 
> > Thanks for your reply.
> 
> I agree with you. The gts of the target data are used during training process. Only the related losses of RPN and RCNN with target domain are discarded when computing the total loss. Something seems inconsistent with the unsupervised domain adaptation object detection declared in the original paper.
> 
> Besides, if the gts of the target domain are unavailable when training, i wonder the generation of the prorotype in target domain.

The author is no longer answering this question...
